crates/services/p2p/src/peer_manager.rs

```diff
@@ -0,0 +1,815 @@
+                                debug!(
+                                    target: "fuel-libp2p",
+                                    "Node {:?} has reported more than {} addresses; it is identified by {:?} and {:?}",
+                                    peer_id, MAX_IDENTIFY_ADDRESSES, protocol_version, agent_version
+                                );
+                                listen_addrs.truncate(MAX_IDENTIFY_ADDRESSES);
+                            }
+
+                            self.peer_manager.insert_peer_info(
+                                &peer_id,
+                                PeerInfoInsert::ClientVersion(agent_version),
+                            );
+                            self.peer_manager.insert_peer_info(
+                                &peer_id,
+                                PeerInfoInsert::Addresses(listen_addrs.clone()),
+                            );
+
+                            let event = PeerInfoEvent::PeerIdentified {
+                                peer_id,
+                                addresses: listen_addrs,
+                            };
+                            return Poll::Ready(NetworkBehaviourAction::GenerateEvent(
+                                event,
+                            ))
+                        }
+                        IdentifyEvent::Error { peer_id, error } => {
+                            debug!(target: "fuel-libp2p", "Identification with peer {:?} failed => {}", peer_id, error)
+                        }
+                        _ => {}
+                    }
+                }
+            }
+        }
+
+        Poll::Pending
+    }
+
+    fn on_connection_handler_event(
+        &mut self,
+        peer_id: PeerId,
+        connection_id: ConnectionId,
+        event: <<Self::ConnectionHandler as IntoConnectionHandler>::Handler as
+            ConnectionHandler>::OutEvent,
+    ) {
+        match event {
+            EitherOutput::First(ping_event) => {
+                self.ping
+                    .on_connection_handler_event(peer_id, connection_id, ping_event)
+            }
+            EitherOutput::Second(identify_event) => self
+                .identify
+                .on_connection_handler_event(peer_id, connection_id, identify_event),
+        }
+    }
+}
+
+// Info about a single Peer that we're connected to
+#[derive(Debug, Default, Clone)]
+pub struct PeerInfo {
+    pub peer_addresses: HashSet<Multiaddr>,
+    pub client_version: Option<String>,
+    pub latest_ping: Option<Duration>,
+}
+
+enum PeerInfoInsert {
+    Addresses(Vec<Multiaddr>),
+    ClientVersion(String),
+    LatestPing(Duration),
+}
+
+/// Manages Peers and their events
+#[derive(Debug, Default, Clone)]
+struct PeerManager {
+    pending_events: VecDeque<PeerInfoEvent>,
+    non_reserved_connected_peers: HashMap<PeerId, PeerInfo>,
+    reserved_connected_peers: HashMap<PeerId, PeerInfo>,
+    reserved_peers: HashSet<PeerId>,
```

Comment:
we should add a counter for the number of reserved peers in our prometheus metrics

cc @ControlCplusControlV